### PR TITLE
Switch dotnet-run to System.CommandLine

### PR DESCRIPTION
**Pull request by krwq**

### CODE REVIEW

**Comment by krwq** [comment]
@anurse @Sridhar-MS @brthor @livarcocc  wanna take a look?

File: src/Microsoft.DotNet.Tools.Compiler/project.json [hunk 1]
```diff
@@ -21,7 +21,8 @@
     "Microsoft.Extensions.CommandLineUtils.Sources": {
       "type": "build",
       "version": "1.0.0-*"
-    }
+    },
+    "System.CommandLine" : "0.1.0-d111815-3"
```
Comment (Sridhar-MS) [inline src/Microsoft.DotNet.Tools.Compiler/project.json]: Seems like you have added only the dependency for dotnet-compile. Do you need this?

File: src/Microsoft.DotNet.Tools.Run/Program.cs [hunk 2]
```diff
@@ -19,52 +13,24 @@ public static int Main(string[] args)
...
-                }
-                return Run(context, configuration.Value() ?? Constants.DefaultConfiguration, app.RemainingArguments, preserveTemporaryOutput.HasValue());
+                syntax.HandleErrors = false;
+                syntax.DefineOption("f|framework", ref runCmd.Framework, "Compile a specific framework");
+                syntax.DefineOption("c|configuration", ref runCmd.Configuration, "Configuration under which to build");
+                syntax.DefineOption("t|preserve-temporary", ref runCmd.PreserveTemporary, "Keep the output's temporary directory around");
+                syntax.DefineOption("p|project", ref runCmd.Project, "The path to the project to run (defaults to the current directory). Can be a path to a project.json or a project directory");
+
+                // TODO: this is not supporting args which can be switches (i.e. --test)
+                // TODO: we need to make a change in System.CommandLine or parse args ourselves.
```
Comment (brthor) [inline src/Microsoft.DotNet.Tools.Run/Program.cs]: Should we create an issue for this? I think this is an important case in the some of the other commands.

File: src/Microsoft.DotNet.Tools.Run/RunCommand.cs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            }
+        }
+
+        private void CalculateDefaultsForNonAssigned()
+        {
+            if (!string.IsNullOrEmpty(Project))
+            {
+                if (File.Exists(Project) && (Path.GetExtension(Project) == ".csx"))
+                {
+                    _isInteractive = true;
```
Comment (brthor) [inline src/Microsoft.DotNet.Tools.Run/RunCommand.cs]: Would this check be more appropriate in the parent function? Doesn't seem to have much to do with calculating defaults.

File: src/Microsoft.DotNet.Tools.Compiler/project.json (hunk 1 above)
Comment (krwq) [inline src/Microsoft.DotNet.Tools.Compiler/project.json]: I'm not sure what do you mean.

Comment (Sridhar-MS) [inline src/Microsoft.DotNet.Tools.Compiler/project.json]: Have you already made changes in Microsoft.DotNet.Tools.Compiler to use System.CommandLine?

File: src/Microsoft.DotNet.Tools.Run/Program.cs (hunk 2 above)
Comment (krwq) [inline src/Microsoft.DotNet.Tools.Run/Program.cs]: https://github.com/dotnet/cli/issues/305

File: src/Microsoft.DotNet.Tools.Run/RunCommand.cs (hunk 3 above)
Comment (krwq) [inline src/Microsoft.DotNet.Tools.Run/RunCommand.cs]: you're right will change

File: src/Microsoft.DotNet.Tools.Run/RunCommand.cs [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+            result = Command.Create(outputName, string.Join(" ", _args))
+                .ForwardStdOut()
+                .ForwardStdErr()
+                .EnvironmentVariable("DOTNET_HOME", runtime)
+                .Execute();
+
+            // Clean up
+            if (!PreserveTemporary)
+            {
+                Directory.Delete(tempDir, recursive: true);
```
Comment (brthor) [inline src/Microsoft.DotNet.Tools.Run/RunCommand.cs]: Could see this throwing for weird reasons (e.g. weird file permissions if directory was already there) Should probably catch this.

Comment (krwq) [inline src/Microsoft.DotNet.Tools.Run/RunCommand.cs]: I'll address this separately (either will create another PR just after this or create an issue)

Comment (brthor) [inline src/Microsoft.DotNet.Tools.Run/RunCommand.cs]: Sounds good :)

**Comment by krwq** [comment]
lol:
```
Caused by: java.io.IOException: remote file operation failed: /mnt/j/workspace/Private/dotnet_cli_release_ubuntu_prtest at hudson.remoting.Channel@15f23ee2:dci-ub-fbld-5: java.io.IOException: No space left on device
```

For the sake of green check I'll rerun

**Comment by krwq** [comment]
@dotnet-bot test this please

**Comment by piotrpMSFT** [comment]
@brthor here we go again...

**Comment by brthor** [comment]
@piotrpMSFT This is a separate issue @Sridhar-MS was helping track down last week. CoreCLR is not cleaning up debug pipes created.

**Comment by piotrpMSFT** [comment]
@sridhar-ms, please link that issue.

**Comment by Sridhar-MS** [comment]
https://github.com/dotnet/coreclr/issues/2126

**Comment by brthor** [comment]
Gone ahead and cleaned up the tmp files in that machine anyways. @Sridhar-MS

**Comment by krwq** [comment]
@dotnet-bot test this please

File: src/Microsoft.DotNet.Tools.Run/Program.cs [hunk 5]
```diff
@@ -19,52 +13,24 @@ public static int Main(string[] args)
...
-                    context = contexts.First();
-                }
-                else
-                {
-                    var fx = NuGetFramework.Parse(framework.Value());
-                    context = contexts.FirstOrDefault(c => c.TargetFramework.Equals(fx));
-                }
-                return Run(context, configuration.Value() ?? Constants.DefaultConfiguration, app.RemainingArguments, preserveTemporaryOutput.HasValue());
+                syntax.HandleErrors = false;
+                syntax.DefineOption("f|framework", ref runCmd.Framework, "Compile a specific framework");
```
Comment (anurse) [inline src/Microsoft.DotNet.Tools.Run/Program.cs]: Oof, I definitely preferred the `var framework = ...` model of capturing these arguments over using `ref`, but I suppose if that's the direction `System.CommandLine` has gone...

Comment (krwq) [inline src/Microsoft.DotNet.Tools.Run/Program.cs]: I actually prefer as is as you can easier set defaults in theory and you can technically implement stuff like (we are not doing that though):
```csharp
List<int> myDefaultList = new List<int>() { 1, 2, 3}; // im not sure what the syntax is
// add to list:
syntax.DefineOption(...., ref myDefaultList, Options.AddToListOrCreateIfNotExists);
```

returning object is quite limiting

**Comment by krwq** [comment]
@dotnet-bot test this please

**Pull request merged**